### ISSUE TITLE
`atomic_condition_variable` might suffer from ABA problem

### ISSUE THREAD

**Issue by zwhfly** [dr-m/atomic_sync#3]
In examples/atomic_condition_variable.h, the `atomic_condition_variable::wait` function does this:
```c++
  template<class mutex> void wait(mutex &m)
  {
    const uint32_t val = fetch_add(1, std::memory_order_acquire);
    m.unlock();
    wait(1 + val);
    m.lock();
  }
```
If the same cv object is used to wait on different conditions, the loop inside `wait(1 + val);` might get waken up and see the same old value and sleep again, causing a deadlock.

Consider this code:
```c++
class single_producer_single_consumer_queue
{
private:
    static constexpr std::size_t N = 8;
    static constexpr std::size_t m_mask = N - 1;
    using data_t = std::array<int, N>;
    data_t m_data;
    atomic_mutex m_mutex;
    atomic_condition_variable m_cv;
    std::size_t m_tail{ 0 };
    std::size_t m_head{ 0 };
public:
    void push(int packet)
    {
        std::unique_lock<atomic_mutex> lock(m_mutex);
        for (;;)
        {
            if ((m_tail - m_head) & m_mask) != 1)
                break;
            m_cv.wait(lock);
        }
        m_data[m_head++] = packet;
        m_head &= m_mask;
        lock.unlock();
        m_cv.signal();//notify_one
    }
    int pull()
    {
        std::unique_lock<atomic_mutex> lock(m_mutex);
        for (;;)
        {
            if (m_head != m_tail)
                break;
            m_cv.wait(lock);
        }
        int r = m_data[m_tail++];
        m_tail &= m_mask;
        lock.unlock();
        m_cv.signal();//notify_one
        return r;
    }
};

single_producer_single_consumer_queue queue;

void thread1()
{
    for (int i = 0; i < 1000; ++i)
        queue.push(i);
    queue.push(-1);
}

void thread2()
{
    int sum = 0;
    for (;;)
    {
        int i = queue.pull();
        if (i == -1)
            break;
        sum += i;
    }
    std::cout << sum << std::endl;
}
```

These events might happen, in this order:
1. queue is empty, thread1 and thread2 start to run
1. thread2 blocks on `queue.pull()`->`m_cv.wait(lock)`->`wait(1)`
1. thread1 pushes one packet, resets the `std::atomic<uint32_t>` inside the `atomic_condition_variable` to 0 and wakes thread2
1. thread2 unblocks; and the thread gets scheduled out by the OS before checking the underlying `std::atomic<uint32_t>`'s value
1. thread1 continues to push packets to the queue, notifies nobody since no threads are blocking on the `std::atomic<uint32_t>`
1. the queue gets full and thread1 blocks on `queue.push(i)`->`m_cv.wait(lock)`->`wait(1)`; the underlying `std::atomic<uint32_t>` is set to 1 before the blocking
1. thread2 continues to run and checks the underlying `std::atomic<uint32_t>`'s value, finds out that it's still 1 and blocks again
1. now everybody is blocking and deadlock occurs

**Comment by dr-m** [comment]
Thank you for your feedback. This could be viewed as a design tradeoff. The condition variable is of minimal storage size, so that it could more easily be placed in the same cache line with the data structures that the condition depends on. If you have multiple conditions, the overhead of introducing a condition variable for each of them is small, only 4 bytes.

Associating multiple conditions with a single condition variable may be a bad idea because of spurious wake-ups of those threads whose condition was not fulfilled yet.

Do you have an idea how `atomic_condition_variable` could be fixed with respect to this scenario, while not significantly increasing its storage size?

On a somewhat related note, I have seen the use of multiple condition variables with one mutex (such as "the buffer is empty" and "the buffer is full" in a producer/consumer system), but not the use of one condition variable with multiple mutexes. The wording in https://linux.die.net/man/3/pthread_cond_wait would explicitly forbid the latter:
>The effect of using more than one mutex for concurrent `pthread_cond_timedwait()` or `pthread_cond_wait()` operations on the same condition variable is undefined; that is, a condition variable becomes bound to a unique mutex when a thread waits on the condition variable, and this (dynamic) binding shall end when the wait returns.

**Comment by FooBarrior** [comment]
Hi all. There is an _event count_ idiom you mihght want to be aware of:
https://github.com/facebook/folly/blob/main/folly/experimental/EventCount.h

Some article behind that, with a semi-proof:
https://pvk.ca/Blog/2019/01/09/preemption-is-gc-for-memory-reordering/

**Comment by dr-m** [comment]
Hi @FooBarrior, I was not aware of that. Thank you for the pointer. Unlike `atomic_shared_mutex` and `atomic_mutex`, the `atomic_condition_variable` is not being used in any real application.

It may take several weeks for me to address this issue. Meanwhile, pull requests are welcome.

**Comment by zwhfly** [comment]
I think there is another problem in the `atomic_condition_variable` class. The `signal` function clears the counter but notifies only one thread. Subsequent `signal` call would not notify remainder waiting threads.

For example, this code gets dead lock:
```c++
class Semaphore
{
    atomic_mutex<> m_mutex;
    atomic_condition_variable m_cv;
    unsigned count{ 0 };
public:
    void put_one()
    {
        std::unique_lock<atomic_mutex<>> lock(m_mutex);
        ++count;
        assert(count != 0);//assume no wrap
        lock.unlock();
        m_cv.signal();//notify_one
    }
    void get_one()
    {
        std::unique_lock<atomic_mutex<>> lock(m_mutex);
        while (count == 0)
            m_cv.wait(lock);
        --count;
        lock.unlock();
    }
};


int main()
{
    Semaphore sem;

    std::thread getter_thread_1{[&]
    {
        std::puts("getter_thread_1 waiting");
        sem.get_one();
        std::puts("getter_thread_1 got one");
    }};
    std::thread getter_thread_2{[&]
    {
        std::puts("getter_thread_2 waiting");
        sem.get_one();
        std::puts("getter_thread_2 got one");
    }};

    using namespace std::chrono_literals;
    std::this_thread::sleep_for(2s);

    std::thread putter_thread{[&]
    {
        sem.put_one();
        sem.put_one();
        std::puts("putter_thread put two");
    }};

    getter_thread_1.join();
    getter_thread_2.join();
    putter_thread.join();
    return 0;
}
```

**Comment by zwhfly** [comment]
> Do you have an idea how `atomic_condition_variable` could be fixed with respect to this scenario, while not significantly increasing its storage size?

This might work. I haven't thought this through though.

* Put this struct `struct { int thread_id; unsigned waiting_count; }` in an atomic.
* `wait()` atomically puts current thread id in `thread_id` and increases `waiting_count` then calls `atomic::wait()`.
* `signal/notify_one()` atomically clears `thread_id` and decreases `waiting_count` then calls `atomic::notify_one()`.
* `broadcast/notify_all()` atomically clears `thread_id` and clears `waiting_count` then calls `atomic::notify_all()`.

Maybe there is a way to split it into two atomics.

**Comment by dr-m** [comment]
Thank you for the second test program.

I believe that both problems could be solved by making the `atomic_condition_variable` similar to `atomic_mutex`: wrap both a flag and a count in a single `std::atomic<uint32_t>`. If the flag is allocated in the most significant bit, the generated code could make use of a sign flag if one is available in the target ISA.

I will need to think if a flag (1-bit counter) is going to be enough or if we would need (for example) two 16-bit counters encoded in the 32-bit integer. Storing a thread identifier should be an overkill. We have `std::thread::id` in `atomic_recursive_shared_mutex`, which has semantics similar to `pthread_mutexattr_settype(attr, PTHREAD_MUTEX_RECURSIVE)`.

**Comment by dr-m** [comment]
@zwhfly I can reproduce the deadlock with the second test program. I added some missing `#include` and compiled it like this:
```diff
diff --git a/test/CMakeLists.txt b/test/CMakeLists.txt
index d508e64..2d65722 100644
--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -1,5 +1,6 @@
 ADD_EXECUTABLE (test_atomic_sync test_atomic_sync.cc)
 ADD_EXECUTABLE (test_atomic_condition test_atomic_condition.cc)
+ADD_EXECUTABLE (test_ac test_ac.cc)
 ADD_EXECUTABLE (test_mutex test_mutex.cc)
 ADD_EXECUTABLE (test_native_mutex test_native_mutex.cc)
 FIND_PACKAGE (Threads)
@@ -28,5 +29,10 @@ TARGET_LINK_LIBRARIES (test_atomic_condition LINK_PUBLIC
   ${ELISION_LIBRARY}
   Threads::Threads)
 
+TARGET_LINK_LIBRARIES (test_ac LINK_PUBLIC
+  atomic_mutex
+  atomic_condition_variable
+  Threads::Threads)
+
 TARGET_LINK_LIBRARIES (test_mutex LINK_PUBLIC atomic_mutex Threads::Threads)
 TARGET_LINK_LIBRARIES (test_native_mutex LINK_PUBLIC atomic_mutex Threads::Threads)
```
Side note: I thought that it would be better practice to signal a condition variable while still holding the mutex, like this:
```c++
    void put_one()
    {
        std::unique_lock<atomic_mutex<>> lock(m_mutex);
        ++count;
        assert(count != 0);//assume no wrap
        m_cv.signal();//notify_one
    }
```
The original variant called `lock.unlock()` before `m_cv.signal()`. If an execution of `get_one()` in another thread got between the mutex release and the `m_cv.signal()`, then that call could cause a spurious wakeup of yet another `get_one()`, because we could have `count==0` again. In any case, this should only be a minor performance problem, not affecting correctness.

I did not come up with a fix yet.

**Comment by dr-m** [comment]
I understood that the "waiters" flag or counter had better be at the least significant end of the lock word, so that when `notify()` or `broadcast()` increments the event count and an overflow occurs, the event count will cleanly wrap around and the "waiters" flag or count will remain intact.
```diff
diff --git a/examples/atomic_condition_variable.h b/examples/atomic_condition_variable.h
index 1e8eae1..a19279f 100644
--- a/examples/atomic_condition_variable.h
+++ b/examples/atomic_condition_variable.h
@@ -73,33 +73,40 @@ public:
 
   template<class mutex> void wait(mutex &m)
   {
-    const uint32_t val = fetch_add(1, std::memory_order_acquire);
+    const uint32_t val = fetch_or(1, std::memory_order_acquire);
     m.unlock();
-    wait(1 + val);
+    wait(val | 1);
     m.lock();
   }
 
   template<class mutex> void wait_shared(mutex &m)
   {
-    const uint32_t val = fetch_add(1, std::memory_order_acquire);
+    const uint32_t val = fetch_or(1, std::memory_order_acquire);
     m.unlock_shared();
-    wait(1 + val);
+    wait(val | 1);
     m.lock_shared();
   }
 
   template<class mutex> void wait_update(mutex &m)
   {
-    const uint32_t val = fetch_add(1, std::memory_order_acquire);
+    const uint32_t val = fetch_or(1, std::memory_order_acquire);
     m.unlock_update();
-    wait(1 + val);
+    wait(val | 1);
     m.lock_update();
   }
 
-  bool is_waiting() const noexcept { return load(std::memory_order_acquire); }
+  bool is_waiting() const noexcept
+  { return load(std::memory_order_acquire) & 1; }
 
   void signal() noexcept
-  { if (exchange(0, std::memory_order_release)) notify_one(); }
+  {
+    uint32_t val = fetch_add(2, std::memory_order_release);
+    if (val & 1)
+    {
+      fetch_and(~1U, std::memory_order_release);
+      notify_all();
+    }
+  }
 
-  void broadcast() noexcept
-  { if (exchange(0, std::memory_order_release)) notify_all(); }
+  void broadcast() noexcept { signal(); }
 };
```
This seems to fix @zwhfly’s second example for me. Having a "waiters" flag instead of a count is not ideal: `signal()` actually needs to `notify_all()` because we do not know how many waiters there are. It could be better to allocate (say) 16 bits for the waiters count and 16 bits for the event count:
```diff
diff --git a/examples/atomic_condition_variable.h b/examples/atomic_condition_variable.h
index 1e8eae1..cd3c009 100644
--- a/examples/atomic_condition_variable.h
+++ b/examples/atomic_condition_variable.h
@@ -62,6 +62,7 @@ class atomic_condition_variable : private std::atomic<uint32_t>
   void notify_all() noexcept { FUTEX(WAKE, INT_MAX); }
   void wait(uint32_t old) const noexcept { FUTEX(WAIT, old); }
 #endif
+  static constexpr uint32_t EVENT = 1U << 16;
 public:
   /** Default constructor */
   constexpr atomic_condition_variable() : std::atomic<uint32_t>(0) {}
@@ -95,11 +96,26 @@ public:
     m.lock_update();
   }
 
-  bool is_waiting() const noexcept { return load(std::memory_order_acquire); }
+  bool is_waiting() const noexcept
+  { return load(std::memory_order_acquire) & (EVENT - 1); }
 
   void signal() noexcept
-  { if (exchange(0, std::memory_order_release)) notify_one(); }
+  {
+    uint32_t val = fetch_add(EVENT, std::memory_order_release);
+    if (val & (EVENT - 1))
+    {
+      fetch_sub(1, std::memory_order_release);
+      notify_one();
+    }
+  }
 
   void broadcast() noexcept
-  { if (exchange(0, std::memory_order_release)) notify_all(); }
+  {
+    uint32_t val = fetch_add(EVENT, std::memory_order_release);
+    if (val & (EVENT - 1))
+    {
+      fetch_sub(1, std::memory_order_release);
+      notify_all();
+    }
+  }
 };
```
I did not carefully consider the `std::memory_order` that I am using above. The second variant also has the benefit that for IA-32 and AMD64, `fetch_add()` and `fetch_sub()` can be translated into the 80486 `lock xadd` instruction. You may be aware that on those ISA, many atomic operations (such as a full-result-returning `fetch_or()`) must be translated into a loop around `lock cmpxchg`, which is extremely inefficient.

At an extreme, it might be that `signal()` and `broadcast()` only need a 1-bit event counter that they would be toggling. That is, we could use 31 bits for counting waiting threads.

@FooBarrior and @zwhfly, what do you think? Would this fix the first test program as well? (I did not try it yet.)

**Comment by dr-m** [comment]
There were some minor issues with the first example program. I edited it a bit, and I was able to reproduce a hang between `single_producer_single_consumer_queue::pull()` and `single_producer_single_consumer_queue::push()`, both waiting in `atomic_condition_variable::wait()` on an object like this:
```
(gdb) p *this
$1 = {m_data = {_M_elems = {24, 17, 18, 19, 20, 21, 22, 23}}, m_mutex = {storage = {m = std::atomic<unsigned int> = { 0 }}}, 
  m_cv = {<std::atomic<unsigned int>> = std::atomic<unsigned int> = { 1 }, <No data fields>}, m_tail = 1, m_head = 1}
```
Here is my revised version of the program:
```c++
#include <cstdio>
#include <thread>
#include <array>
#include <cassert>
#include <mutex>
#include "atomic_mutex.h"
#include "atomic_condition_variable.h"

class single_producer_single_consumer_queue
{
private:
    static constexpr std::size_t N = 8;
    static constexpr std::size_t m_mask = N - 1;
    using data_t = std::array<int, N>;
    data_t m_data;
    atomic_mutex<> m_mutex;
    atomic_condition_variable m_cv;
    std::size_t m_tail{ 0 };
    std::size_t m_head{ 0 };
public:
    void push(int packet)
    {
        std::unique_lock<atomic_mutex<>> lock(m_mutex);
        for (;;)
        {
            if (((m_tail - m_head) & m_mask) != 1)
                break;
            m_cv.wait(lock);
        }
        m_data[m_head++] = packet;
        m_head &= m_mask;
        m_cv.signal();//notify_one
    }
    int pull()
    {
        std::unique_lock<atomic_mutex<>> lock(m_mutex);
        for (;;)
        {
            if (m_head != m_tail)
                break;
            m_cv.wait(lock);
        }
        int r = m_data[m_tail++];
        m_tail &= m_mask;
        m_cv.signal();//notify_one
        return r;
    }
};

int main()
{
    single_producer_single_consumer_queue queue;

    std::thread producer{[&]
    {
        for (int i = 0; i < 1000; ++i) queue.push(i);
        queue.push(-1);
        std::puts("produced");
    }};

    int sum = 0;
    for (;;)
    {
        int i = queue.pull();
        if (i == -1)
            break;
        sum += i;
    }
    std::printf("consumed %d\n", sum);
    producer.join();
    return 0;
}
```
With the latter patch of my above comment, the program does not hang for me, but it each time outputs the correct sum:
```
produced
consumed 499500
```
I suspect that if I make `EVENT = 1U<<31` or even `EVENT = 1U << 8`, the chances of hitting the original problem (missed wakeup) could be too probable, and we’d better always call `notify_all()` instead of `notify_one()`. The 16+16 split could be something that works reasonably well in practice, at least until massively parallel systems with tens of thousands of CPU cores become available.

I would greatly appreciate your feedback or testing of this.

**Comment by dr-m** [comment]
For what it is worth, even with the following change both test programs completed without a hang on my system, for the few times I ran them:
```diff
diff --git a/examples/atomic_condition_variable.h b/examples/atomic_condition_variable.h
index cd3c009..89cbed7 100644
--- a/examples/atomic_condition_variable.h
+++ b/examples/atomic_condition_variable.h
@@ -62,7 +62,7 @@ class atomic_condition_variable : private std::atomic<uint32_t>
   void notify_all() noexcept { FUTEX(WAKE, INT_MAX); }
   void wait(uint32_t old) const noexcept { FUTEX(WAIT, old); }
 #endif
-  static constexpr uint32_t EVENT = 1U << 16;
+  static constexpr uint32_t EVENT = 1U << 31;
 public:
   /** Default constructor */
   constexpr atomic_condition_variable() : std::atomic<uint32_t>(0) {}
```
Furthermore, I tested a variant of the program in my previous comment with 3 producer threads (each producing 1000 values) and 1 consumer (processing 3000 elements) and it did not hang either.

I’d appreciate a test program that would use a condition variable in N threads for some M steps (both being run-time parameters). That would convince better of the correctness. The scenario that I fear would be that in one blocked `wait()` the expected `1 + val` will be reused, because the event count had meanwhile done a ‘full cycle’ (including a wrap-around) and some other threads had finished and entered `atomic_condition_variable::wait()`.

**Comment by dr-m** [comment]
In my fix, I updated `test_atomic_condtiion` so that it would reproduce a hang. While working on it, I noticed that a reliable operation of `atomic_condition_variable::is_waiting()` requires that the pending wait count be both incremented and decremented in `atomic_condition_variable::wait()`.
